chains/solana/contracts/programs/cctp-token-pool/src/lib.rs

```diff
@@ -0,0 +1,714 @@
+    }
+
+    pub fn edit_chain_remote_config_cctp(
+        ctx: Context<EditChainConfig>,
+        _remote_chain_selector: u64,
+        _mint: Pubkey,
+        cfg: CctpChain,
+    ) -> Result<()> {
+        ctx.accounts.chain_config.cctp = cfg;
+        emit!(RemoteChainCctpConfigChanged {
+            config: ctx.accounts.chain_config.cctp
+        });
+        Ok(())
+    }
+
+    // Add remote pool addresses
+    pub fn append_remote_pool_addresses(
+        ctx: Context<AppendRemotePoolAddresses>,
+        remote_chain_selector: u64,
+        _mint: Pubkey,
+        addresses: Vec<RemoteAddress>,
+    ) -> Result<()> {
+        ctx.accounts.chain_config.base.append_remote_pool_addresses(
+            remote_chain_selector,
+            _mint,
+            addresses,
+        )
+    }
+
+    // set rate limit
+    pub fn set_chain_rate_limit(
+        ctx: Context<EditChainConfig>,
+        remote_chain_selector: u64,
+        mint: Pubkey,
+        inbound: RateLimitConfig,
+        outbound: RateLimitConfig,
+    ) -> Result<()> {
+        ctx.accounts.chain_config.base.set_chain_rate_limit(
+            remote_chain_selector,
+            mint,
+            inbound,
+            outbound,
+        )
+    }
+
+    // delete chain config
+    pub fn delete_chain_config(
+        _ctx: Context<DeleteChainConfig>,
+        remote_chain_selector: u64,
+        mint: Pubkey,
+    ) -> Result<()> {
+        emit!(RemoteChainRemoved {
+            chain_selector: remote_chain_selector,
+            mint
+        });
+        Ok(())
+    }
+
+    pub fn configure_allow_list(
+        ctx: Context<AddToAllowList>,
+        add: Vec<Pubkey>,
+        enabled: bool,
+    ) -> Result<()> {
+        ctx.accounts.state.config.list_enabled = enabled;
+        let list = &mut ctx.accounts.state.config.allow_list;
+        for key in add {
+            require!(
+                !list.contains(&key),
+                CcipTokenPoolError::AllowlistKeyAlreadyExisted
+            );
+            list.push(key);
+        }
+
+        Ok(())
+    }
+
+    pub fn remove_from_allow_list(
+        ctx: Context<RemoveFromAllowlist>,
+        remove: Vec<Pubkey>,
+    ) -> Result<()> {
+        let list = &mut ctx.accounts.state.config.allow_list;
+        for key in remove {
+            require!(
+                list.contains(&key),
+                CcipTokenPoolError::AllowlistKeyDidNotExist
+            );
+            list.retain(|k| k != &key);
+        }
+
+        Ok(())
+    }
+
+    pub fn release_or_mint_tokens<'info>(
+        ctx: Context<'_, '_, 'info, 'info, TokenOfframp<'info>>,
+        release_or_mint: ReleaseOrMintInV1,
+    ) -> Result<ReleaseOrMintOutV1> {
+        require!(
+            !release_or_mint.offchain_token_data.is_empty(),
+            CctpTokenPoolError::InvalidTokenData
+        );
+
+        let msg_att = MessageAndAttestation::try_from_slice(&release_or_mint.offchain_token_data)
```

Comment:
This conversion is duplicated later inside `parse_remaining_release_accounts` in order to retrieve the nonce seed. It would be good, if easy, to make the conversion only once here and pass down `&MessageAndAttestation`, to save some cycles.

Reply:
You're right, I've changed it now :+1: